packages/dd-trace/src/appsec/iast/taint-tracking/rewriter.js

```diff
@@ -114,11 +138,70 @@ function enableRewriter (telemetryVerbosity) {
       }
       shimmer.wrap(Module.prototype, '_compile', compileMethod => getCompileMethodFn(compileMethod))
     }
+
+    enableEsmRewriter(telemetryVerbosity)
   } catch (e) {
     log.error('[ASM] Error enabling TaintTracking Rewriter', e)
   }
 }
 
+function isEsmConfigured () {
+  const hasLoaderArg = !!process.execArgv?.find(
+    arg => arg.includes('--loader') || arg.includes('--experimental-loader')
+  )
+  if (hasLoaderArg) return true
+
+  const nodeOptions = process.env.NODE_OPTIONS
+  const hasLoaderNodeOptions = nodeOptions &&
+    (nodeOptions.includes('--loader') || nodeOptions.includes('--experimental-loader'))
+  if (hasLoaderNodeOptions) return true
+
+  const initializeLoaded = Object.keys(require.cache).find(file => file.includes('import-in-the-middle/hook.js'))
+  return !!initializeLoaded
+}
+
+function enableEsmRewriter (telemetryVerbosity) {
+  if (isMainThread && Module.register && !esmRewriterEnabled && isEsmConfigured()) {
+    esmRewriterEnabled = true
+
+    const { port1, port2 } = new MessageChannel()
+
+    port1.on('message', (message) => {
+      const { type, data } = message
+      switch (type) {
+        case LOG:
+          log[data.level]?.(...data.messages)
+          break
+
+        case REWRITTEN:
+          esmRewritePostProcess(data.rewritten, data.url)
+          break
+      }
+    })
+
+    port1.unref()
+    port2.unref()
+
+    const chainSourceMap = isFlagPresent('--enable-source-maps')
+    const data = {
+      port: port2,
+      csiMethods,
+      telemetryVerbosity,
+      chainSourceMap
+    }
+
+    try {
+      Module.register('./rewriter-esm.mjs', {
+        parentURL: pathToFileURL(__filename),
+        transferList: [port2],
+        data
+      })
+    } catch (e) {
+      log.error('[ASM] Error enabling ESM Rewriter', e)
```

Comment:
should we call `port1.off`?

Reply:
Good point, I checked the api, and I think that is better if we call to `port1.close()` and `port2.close()` methods. changed.